src/main/scala/com/raquo/airstream/web/AjaxEventStream.scala

```diff
@@ -0,0 +1,162 @@
+package com.raquo.airstream.web
+
+import com.raquo.airstream.core.Transaction
+import com.raquo.airstream.eventstream.EventStream
+import org.scalajs.dom
+
+/**
+  * [[AjaxEventStream]] performs an HTTP request and emits the [[dom.XMLHttpRequest]]/[[dom.ext.AjaxException]] on
+  * success/failure.
+  *
+  * The network request is delayed until start. In other words, the network request is not performed if the stream is
+  * never started.
+  *
+  * On restart, a new request is performed and the subsequent success/failure is propagated downstream.
+  */
+class AjaxEventStream(
+  method: String,
+  url: String,
+  data: dom.ext.Ajax.InputData,
+  timeout: Int,
+  headers: Map[String, String],
+  withCredentials: Boolean,
+  responseType: String
```

Comment:
Can we also have a progress observer here?

Something like this:
 
```scala
maybeProgressObserver: Option[Observer[ProgressEvent]]
```

and below
```scala
    maybeProgressObserver.foreach { progressObserver =>
      req.upload.onprogress = p => progressObserver.onNext(p)
    }
```

And add this param to the `apply`/`post`/`put` etc.

There are other events as well:

```scala

  /**
   * A function that is called periodically with information about the progress of the
   * request.
   *
   * MDN
   */
  var onprogress: js.Function1[ProgressEvent, _] = js.native

  /**
   * The function to call when a request encounters an error.
   *
   * MDN
   */
  var onerror: js.Function1[ErrorEvent, _] = js.native

  /**
   * The function to call when an HTTP request returns after successfully loading
   * content.
   *
   * MDN
   */
  var onload: js.Function1[js.Any, _] = js.native

  /**
   * A function that is called if the event times out; this only happens if a timeout has
   * been previously established by setting the value of the XMLHttpRequest object's
   * timeout attribute.
   *
   * MDN
   */
  var ontimeout: js.Function1[js.Any, _] = js.native

  /**
   * The function to call when a request is aborted.
   *
   * MDN
   */
  var onabort: js.Function1[js.Any, _] = js.native

  /**
   * A function that gets called when the HTTP request first begins loading data.
   *
   * MDN
   */
  var onloadstart: js.Function1[js.Any, _] = js.native

  /**
   * A function that is called when the load is completed, even if the request failed.
   *
   * MDN
   */
  var onloadend: js.Function1[ProgressEvent, _] = js.native
```

But so far I've only ever needed the `onprogress`.

Reply:
I'm working on this, it got deeper (but will end up better).